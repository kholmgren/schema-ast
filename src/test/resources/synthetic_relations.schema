definition tenant/user {}

definition tenant/portfolio {
	relation reader: tenant/user
	permission read = reader
}

definition tenant/folder {
	relation parent_portfolio: tenant/portfolio
	relation reader: tenant/user
	permission read = reader + parent_portfolio->read
}

definition tenant/document {
	relation parent_folder: tenant/folder
	relation reader: tenant/user

	/** read defines whether a tenant/user can read the tenant/document */
	permission read = reader + parent_folder->read
}