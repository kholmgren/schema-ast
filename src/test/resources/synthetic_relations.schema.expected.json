{
  "@type" : "schema",
  "definitions" : [ {
    "@type" : "definition",
    "name" : "tenant/user"
  }, {
    "@type" : "definition",
    "name" : "tenant/portfolio",
    "relations" : [ {
      "@type" : "relation",
      "name" : "reader",
      "expression" : {
        "@type" : "ref",
        "ref" : "tenant/user"
      }
    } ],
    "permissions" : [ {
      "@type" : "permission",
      "name" : "read",
      "expression" : {
        "@type" : "ref",
        "ref" : "reader"
      }
    } ]
  }, {
    "@type" : "definition",
    "name" : "tenant/folder",
    "relations" : [ {
      "@type" : "relation",
      "name" : "parent_portfolio",
      "expression" : {
        "@type" : "ref",
        "ref" : "tenant/portfolio"
      }
    }, {
      "@type" : "relation",
      "name" : "reader",
      "expression" : {
        "@type" : "ref",
        "ref" : "tenant/user"
      }
    } ],
    "permissions" : [ {
      "@type" : "permission",
      "name" : "read",
      "expression" : {
        "@type" : "union",
        "left" : {
          "@type" : "ref",
          "ref" : "reader"
        },
        "right" : {
          "@type" : "walk",
          "refs" : [ "parent_portfolio", "read" ]
        }
      }
    } ]
  }, {
    "@type" : "definition",
    "name" : "tenant/document",
    "relations" : [ {
      "@type" : "relation",
      "name" : "parent_folder",
      "expression" : {
        "@type" : "ref",
        "ref" : "tenant/folder"
      }
    }, {
      "@type" : "relation",
      "name" : "reader",
      "expression" : {
        "@type" : "ref",
        "ref" : "tenant/user"
      }
    } ],
    "permissions" : [ {
      "@type" : "permission",
      "name" : "read",
      "expression" : {
        "@type" : "union",
        "left" : {
          "@type" : "ref",
          "ref" : "reader"
        },
        "right" : {
          "@type" : "walk",
          "refs" : [ "parent_folder", "read" ]
        }
      }
    } ]
  } ]
}